src/gluonts/model/deepar/_schema.py

```diff
@@ -0,0 +1,161 @@
+        if not isinstance(other, NumpyArrayField):
+            return False
+        return self.dtype == other.dtype and self.ndim == other.ndim
+
+    def __repr__(self):
+        return f"NumpyArrayField(dtype={self.dtype!r}, ndim={self.ndim!r}"
+
+    def __call__(self, value: Any) -> np.ndarray:
+        value = np.asarray(value, dtype=object)
+        value = value.astype(self.dtype)
+        if self.ndim is not None and self.ndim != value.ndim:
+            raise GluonTSDataError(
+                f"expected array with dimension {self.ndim}, but got {value.ndim}."
+            )
+        return value
+
+    def is_compatible(self, value: Any) -> bool:
+        if not isinstance(value, (list, tuple, np.ndarray)):
+            return False
+
+        try:
+            x = np.asarray(value)
+        except:
+            return False
+
+        # int types
+        if self.dtype in [np.int32, np.int64]:
+            if x.dtype.kind != "i":
+                return False
+            return self.ndim is None or x.ndim == self.ndim
+
+        # float types
+        try:
+            x = np.asarray(value, dtype=self.dtype)
+        except:
+            return False
+        return self.ndim is None or x.ndim == self.ndim
+
+
+Freq = Union[str, pd.DateOffset]
+
+
+class PandasPeriodField(FieldType[pd.Period]):
+    def __init__(
+        self,
+        freq: Freq,
+    ) -> None:
+        self.freq = to_offset(freq) if isinstance(freq, str) else freq
+
+    def __eq__(self, other):
+        if not isinstance(other, PandasPeriodField):
+            return False
+        return self.freq == other.freq
+
+    def __repr__(self):
+        return f"PandasPeriodField(freq={self.freq!r})"
+
+    def __call__(self, value: Any) -> pd.Period:
+        period = PandasPeriodField._process(value, self.freq)
+        return period
+
+    @staticmethod
+    @lru_cache(maxsize=10000)
+    def _process(string: str, freq: pd.DateOffset) -> pd.Timestamp:
+        return pd.Period(string, freq=freq)
+
+    def is_compatible(self, v: Any) -> bool:
+        if not isinstance(v, (str, pd.Period)):
+            return False
+        try:
+            self(v)
+        except:
+            return False
+        return True
+
+
+class Schema:
+    def __init__(self, fields: Dict[str, FieldType]) -> None:
+        self.fields = fields
+        # select fields that should be handled in the loop, because
+        # - they are non optional
+        # - or they are not AnyFields
+        self._fields_for_processing = {k: f for k, f in self.fields.items()}
+
+    def __eq__(self, other):
+        if self.fields.keys() != other.fields.keys():
+            return False
+        return all(self.fields[k] == other.fields[k] for k in self.fields)
+
+    def __repr__(self):
+        return (
+            "Schema(fields={"
+            + ", ".join(f"'{k}':{v}" for k, v in self.fields.items())
+            + "})"
+        )
+
+    def __call__(self, d: Dict[str, Any]) -> None:
```

Comment:
Maybe we can have an `in_place` parameter, which allows us to either mutate the passed value or otherwise returns a copy.